sympy/stats/drv.py

```diff
@@ -30,24 +31,29 @@ def check(*args):
     def sample(self):
         """ A random realization from the distribution """
         icdf = self._inverse_cdf_expression()
-        return floor(icdf(random.uniform(0, 1)))
+        while(True):
+            sample_ = floor(list(icdf(random.uniform(0, 1)))[0])
+            if sample_ >= self.set.inf:
+                return sample_
 
     @cacheit
     def _inverse_cdf_expression(self):
         """ Inverse of the CDF
 
         Used by sample
         """
-        x, z = symbols('x, z', real=True, positive=True, cls=Dummy)
+        x = symbols('x', real = True, positive = True,
+         integer = True, cls=Dummy)
+        y, z = symbols('y z', real=True, positive=True, cls=Dummy)
+        cdf_temp = self.cdf(x).subs(x, y)
```

Comment:
Why not `self.cdf(y)` here, which should be the same? 
And then I don't see x used anymore in this function.  
Also, positive=True implies real=True, so real=True is redundant here.

Reply:
Here, the original CDF function is piecewise and non-invertible. Using `x` which is specified to be an integer isolates the invertible part of the CDF, but then `solveset` is unable to invert it, because of the constraint. For this reason, `y`, which is declared as an integer substitutes `x` , and using `floor` in `sample` guaranteed that the result is an integer.